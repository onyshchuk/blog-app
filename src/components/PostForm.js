import React from 'react';
import moment from 'moment';
import RemoveApprovalModal from './RemoveApprovalModal';

class PostForm extends React.Component {
   state = {
      id: this.props.post ? this.props.post.id : '',
      title: this.props.post ? this.props.post.title : '',
      body: this.props.post ? this.props.post.body : '',
      createdAt: this.props.post ? this.props.post.createdAt : moment().valueOf(),
      error: '',
      isOpen: false
   };
   onTitleChange = e => {
      const title = e.target.value;
      this.setState(() => ({ title }));
   };
   onBodyChange = e => {
      const body = e.target.value;
      this.setState(() => ({ body }));
   };
   onSaveClick = e => {
      e.preventDefault();

      if(this.state.title && this.state.body){
         this.props.handleSaveClick({
            id: this.state.id,
            title: this.state.title,
            body: this.state.body,
            createdAt: this.state.createdAt
         });
         this.setState({ error: '' });
      } else {
         this.setState({ error: 'Pls fill the fields' });
      }
   };
   onRemoveClick = () => {
      this.setState({ isOpen: true });
   };
   closeModal = () => {
      this.setState({ isOpen: false });
   };
   render() {
      return (
         <form className="form" onSubmit={this.onSaveClick}>
            {this.state.error && <p className="form__error">{this.state.error}</p>}
            <input
               className="text-input"
               autoFocus
               type="text"
               value={this.state.title}
               placeholder="Post title" 
               onChange={this.onTitleChange}
            />
            <textarea
               className="textarea"
               value={this.state.body}
               placeholder="Post body" 
               onChange={this.onBodyChange}
            />
            <div>
               <button className="button" type="submit">Save Post</button>
               {this.props.handleRemoveClick && <a className="button" onClick={this.onRemoveClick}>Remove Post</a>}
            </div>
            <RemoveApprovalModal
               isOpen={this.state.isOpen}
               selectedPostTitle={this.state.title}
               handleRemoveClick={this.props.handleRemoveClick}
               closeModal={this.closeModal}
            />
         </form>
      );
   };
};

export default PostForm;