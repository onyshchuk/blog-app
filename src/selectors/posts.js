export default (posts, { text, sortBy } ) => {
   return posts.filter(post => post.title.includes(text))
               .sort((a, b) => {
                  if(sortBy === 'title') 
                     return a.title < b.title ? -1 : a.title > b.title ? 1 : 0;
                  if(sortBy === 'date') 
                     return a.createdAt < b.createdAt ? 1 : a.createdAt > b.createdAt ? -1 : 0;
               });
}