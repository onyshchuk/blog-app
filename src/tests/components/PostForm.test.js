import React from 'react';
import { shallow } from 'enzyme';
import PostForm from './../../components/PostForm';
import posts from './../fixtures/posts';

test('should render PostFrom for AddPage with default values properly', () => {
   const wrapper = shallow(<PostForm />);
   expect(wrapper).toMatchSnapshot();
});

test('should render PostForm for EditPage with post data properly with remove button', () => {
   const wrapper = shallow(<PostForm handleRemoveClick={() => {}} post={posts[2]} />);
   expect(wrapper).toMatchSnapshot();
});

test('should handle Title change properly', () => {
   const wrapper = shallow(<PostForm />);
   const value = 'new title';
   wrapper.find('input').simulate('change', { target: { value } });
   expect(wrapper.state('title')).toBe(value);
});

test('should handle Body change properly', () => {
   const wrapper = shallow(<PostForm />);
   const value = 'new body';
   wrapper.find('textarea').simulate('change', { target: { value } });
   expect(wrapper.state('body')).toBe(value);
});

test('should handle form submition properly', () => {
   const handleSaveClick = jest.fn();
   const wrapper = shallow(<PostForm post={posts[2]} handleSaveClick={handleSaveClick}/>);
   wrapper.find('form').simulate('submit', { preventDefault: () => { } });
   expect(handleSaveClick).toHaveBeenLastCalledWith(posts[2])
});

test('should handle form submition error properly', () => {
   const handleSaveClick = jest.fn();
   const wrapper = shallow(<PostForm handleSaveClick={handleSaveClick} />);
   wrapper.find('form').simulate('submit', { preventDefault: () => { } });
   expect(wrapper.state('error').length).toBeGreaterThan(0);
   expect(wrapper).toMatchSnapshot();
});

test('should handle Remove click properly', () => {
   const handleRemoveClick = jest.fn();
   const wrapper = shallow(<PostForm post={posts[2]} handleRemoveClick={handleRemoveClick} />);
   wrapper.find('a').simulate('click');
   expect(wrapper.state('isOpen')).toBe(true);
});

test('should handle close modal properly', () => {
   const wrapper = shallow(<PostForm />);
   wrapper.setState({ isOpen: true });
   wrapper.find('RemoveApprovalModal').prop('closeModal')();
   expect(wrapper.state('isOpen')).toBe(false);
});