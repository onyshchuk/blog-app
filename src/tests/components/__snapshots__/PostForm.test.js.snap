// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should handle form submition error properly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <p
    className="form__error"
  >
    Pls fill the fields
  </p>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Post body"
    value=""
  />
  <div>
    <button
      className="button"
      type="submit"
    >
      Save Post
    </button>
  </div>
  <RemoveApprovalModal
    closeModal={[Function]}
    isOpen={false}
    selectedPostTitle=""
  />
</form>
`;

exports[`should render PostForm for EditPage with post data properly with remove button 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value="third post"
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Post body"
    value="third body"
  />
  <div>
    <button
      className="button"
      type="submit"
    >
      Save Post
    </button>
    <a
      className="button"
      onClick={[Function]}
    >
      Remove Post
    </a>
  </div>
  <RemoveApprovalModal
    closeModal={[Function]}
    handleRemoveClick={[Function]}
    isOpen={false}
    selectedPostTitle="third post"
  />
</form>
`;

exports[`should render PostFrom for AddPage with default values properly 1`] = `
<form
  className="form"
  onSubmit={[Function]}
>
  <input
    autoFocus={true}
    className="text-input"
    onChange={[Function]}
    placeholder="Post title"
    type="text"
    value=""
  />
  <textarea
    className="textarea"
    onChange={[Function]}
    placeholder="Post body"
    value=""
  />
  <div>
    <button
      className="button"
      type="submit"
    >
      Save Post
    </button>
  </div>
  <RemoveApprovalModal
    closeModal={[Function]}
    isOpen={false}
    selectedPostTitle=""
  />
</form>
`;
